### ISSUE TITLE
When output is provided as a dictionary a loss function is not set during compile.

### ISSUE THREAD

**Issue by odinsbane** [keras-team/keras#18959]
Here is a very short example that demonstrates the issue:

```
import keras
import numpy

x = keras.layers.Input((16, 16))
x2 = keras.layers.Dense(32)(x)
x2 = keras.layers.Dense(1, activation = "sigmoid")(x2)

model = keras.models.Model(inputs =[ x], outputs = {"dt":x2})

model.compile( optimizer = keras.optimizers.Adam(0.0001), loss="mean_squared_error")

i = numpy.random.random((5, 16, 16))
o = (numpy.random.random((5, 1))>0.5)*1
model.fit(i, o)
```

Results in

> ValueError: No loss to compute. Provide a `loss` argument in `compile()`.

I don't know how to fix this issue, I tried using `loss = {"dt":"mean_squared_error"} which is how I previously would setup different loss functions for different outputs. Maybe this is a bug, or it is a new way to setup outputs/loss functions.

pip freeze returns keras and tensorflow values:

>keras==3.0.0  
tensorboard==2.15.1
tensorboard-data-server==0.7.2
tensorflow==2.15.0.post1
tensorflow-estimator==2.15.0
tensorflow-io-gcs-filesystem==0.34.0

**Comment by sachinprasadhs** [comment]
Hi, You are providing wrong inputs and outputs.

Please find the below updated code, which is working fine.

```python
import keras
import numpy

x = keras.layers.Input((16, 16))
x2 = keras.layers.Dense(32)(x)
x2 = keras.layers.Dense(1, activation = "sigmoid")(x2)

model = keras.models.Model(inputs=x, outputs=x2)

model.compile( optimizer = keras.optimizers.Adam(0.0001), loss="mean_squared_error")

i = numpy.random.random((5, 16, 16))
o = (numpy.random.random((5, 1))>0.5)*1
model.fit(i, o)
```

**Comment by odinsbane** [comment]
Your response seems to miss the point. Previously the outputs could be
provided as a dictionary. Now it silently fails.

Are you saying that lists and dictionaries are no longer valid arguments
for inputs and outputs? That seems like quite a change to the api and there
should be an error or warning.

Also, what is the replacement? I have multiple outputs, I map them with a
dictionary. What should I do now.

Your example does not solve the problem.

 Thank you

**Comment by sachinprasadhs** [comment]
It should ideally support dictionary inputs, but to only one level, Keras 3 doesn't support nested level input, output dictionary mapping anymore.
For more details, check the migration guide here https://keras.io/guides/migrating_to_keras_3/#using-deeply-nested-inputs-in-functional-models

**Comment by odinsbane** [comment]
That doesn't sound correct because I am not using a "deeply nested"
dictionary and their examples give warning or errors.

**Comment by odinsbane** [comment]
From the documentation for keras 3 model api https://keras.io/api/models/model/#model-class :

>Note: Only dicts, lists, and tuples of input tensors are supported. Nested inputs are not supported (e.g. lists of list or dicts of dict).

So I am providing a dict it should be supported. I would say this is a bug, or the documentation needs to be updated.

**Comment by divyashreepathihalli** [comment]
the behavior change has been documented [here](https://keras.io/guides/migrating_to_keras_3/#:~:text=None%2C%20name%3Dkeras_tensor_6%3E-,Multi%2Doutput%20model%20evaluate(),-The%20evaluate())

**Comment by odinsbane** [comment]
That describes the loss function, it doesn't say that the model no longer
accepts a dictionary. I linked the documentation that says the model
accepts lists, and dictionaries. Maybe you should fix that too.

**Comment by odinsbane** [comment]
This makes keras incompatible with previously trained models for no reason. It also makes it possible to build models and then be unable to train them with a cryptic error. It is really strange to keep a part of the API that is fundamentally broken.

**Comment by odinsbane** [comment]
The documentation still says that the Model API supports dictionaries. https://keras.io/api/models/model_training_apis/

**Comment by ashindea** [comment]
What is the rationale for this? It seems to be a completely unnecessary burden on the users of Keras. If Model.output does not need to handle Maps but only a list, why not have the Model API handle that behind the scenes by converting the Map.values to a List?

**Comment by ghsanti** [comment]
@odinsbane @ashindea trying to help here (just another user.)

I think what you were asking was fixed in recent versions.

Use keras nightly `pip install keras-nightly` and your code should work.

[Here is a Gist.](https://colab.research.google.com/gist/ghsanti/738fb6db4d16d96893a53edc0c3480c4/untitled4.ipynb)
I review part of your original post below.


--------


> Are you saying that lists and dictionaries are no longer valid arguments for inputs and outputs

No, should not be the case, that was a bug.

In the lines below:
> `model = keras.models.Model(inputs =[ x], outputs = {"dt":x2})`
> `model.fit(i, o)`

For best practice `o` should be `{'dt':o}`, but it as it's written it'd also work now.

**Comment by odinsbane** [comment]
@ghsanti wow good find! Thank you for the update. Do you know if it was another bug that go fixed that also fixed this one?

**Comment by ghsanti** [comment]
> @ghsanti wow good find! Thank you for the update. Do you know if it was another bug that go fixed that also fixed this one?

It was fixed between 3.4.1 and current head, so [3.4.1...master](https://github.com/keras-team/keras/compare/v3.4.1...master) should contain it, but I don't know which commit.

You should be able to stick to 3.4.2 once it's released (hopefully this month.)

**Comment by ni-metre** [comment]
Thanks. FWIW I think it's working on the version of keras in tf-keras-nightly 2.18.0.dev2024080509, which contains tf.keras 3.4.1.dev2024080503.
